File: algorithms/sorting/quick-sort/swift/QuickSort.swift
class QuickSort {
    static func sort(_ arr: inout [Int]) {
        if !arr.isEmpty {
            quickSort(&arr, 0, arr.count - 1)
        }
    }

    private static func quickSort(_ arr: inout [Int], _ low: Int, _ high: Int) {
        if low < high {
            let pi = partition(&arr, low, high)

            quickSort(&arr, low, pi - 1)
            quickSort(&arr, pi + 1, high)
        }
    }

    private static func partition(_ arr: inout [Int], _ low: Int, _ high: Int) -> Int {
        let pivot = arr[high]
        var i = (low - 1)

        for j in low..<high {
            if arr[j] < pivot {
                i += 1
                arr.swapAt(i, j)
            }
        }
        arr.swapAt(i + 1, high)
        return i + 1
    }
}

func quickSort(_ arr: [Int]) -> [Int] {
    var result = arr
    QuickSort.sort(&result)
    return result
}

File: algorithms/sorting/quick-sort/rust/quick_sort.rs
pub fn quick_sort(arr: &mut [i32]) {
    if !arr.is_empty() {
        let len = arr.len();
        _quick_sort(arr, 0, (len - 1) as isize);
    }
}

fn _quick_sort(arr: &mut [i32], low: isize, high: isize) {
    if low < high {
        let pi = partition(arr, low, high);
        
        _quick_sort(arr, low, pi - 1);
        _quick_sort(arr, pi + 1, high);
    }
}

fn partition(arr: &mut [i32], low: isize, high: isize) -> isize {
    let pivot = arr[high as usize];
    let mut i = low - 1;
    
    for j in low..high {
        if arr[j as usize] < pivot {
            i += 1;
            arr.swap(i as usize, j as usize);
        }
    }
    arr.swap((i + 1) as usize, high as usize);
    return i + 1;
}
